compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp

```diff
@@ -0,0 +1,931 @@
+                           MLIRContext *ctx)
+      : DataFlowAnalysis(solver), anchors(anchors), ctx(ctx) {}
+
+  LogicalResult initialize(Operation *root) override;
+
+  LogicalResult visit(ProgramPoint point) override;
+
+  /// Register a new value to be part of the dataflow analysis. The value should
+  /// not be part of the analysis already. This is used for new values that are
+  /// created.
+  void registerNewValue(Value val, const VectorLayoutInterface &layout);
+
+  friend class DistributionLayout;
+
+private:
+  void visitOperation(Operation *op);
+
+  void visitRegionSuccessors(RegionBranchOpInterface branch,
+                             RegionBranchPoint branchPoint,
+                             OperandRange operands);
+
+  DistributionLayout *getLatticeElement(Value val);
+
+  DenseMap<Value, VectorLayoutInterface> anchors;
+
+  MLIRContext *ctx;
+};
+
+/// ==========================================================================
+///        DistributionLayout
+/// ==========================================================================
+
+DistributionLayout::ResolutionResult
+DistributionLayout::doResolution(const VectorLayoutInterface &rhs) {
+  VectorLayoutInterface &lhs = vectorLayout;
+
+  // Ignore if the layout to resolve with is empty.
+  if (!rhs) {
+    return ResolutionResult::NoChange;
+  }
+
+  // If both layouts are same, do nothing.
+  if (lhs == rhs) {
+    return ResolutionResult::NoChange;
+  }
+
+  // Take the other layout if the current layout is empty.
+  if (!lhs && rhs) {
+    setInnerLayout(rhs);
+    return ResolutionResult::Change;
+  }
+
+  // Layouts conflict and need to be resolved.
+  return ResolutionResult::Conflict;
+}
+
+ChangeResult DistributionLayout::resolveWithPossibleConflict(
+    const VectorLayoutInterface &rhs, OpOperand &opOperand) {
+  ResolutionResult result = doResolution(rhs);
+
+  // If there is no conflict, simply return.
+  if (result == ResolutionResult::NoChange) {
+    return ChangeResult::NoChange;
+  } else if (result == ResolutionResult::Change) {
+    return ChangeResult::Change;
+  }
+
+  // Resolve conflict by create an operation that takes the input the conflicted
+  // value and returns the resolved value.
+  OpBuilder builder(opOperand.getOwner());
+  Value input = opOperand.get();
+  // Create a resolution operation. This conflict should be handeled later by
+  // someone else, not this analysis.
+  Operation *resolveOp =
+      builder.create<IREE ::VectorExt::LayoutConflictResolutionOp>(
+          input.getLoc(), input.getType(), input, vectorLayout, rhs);
+  Value resolvedValue = resolveOp->getResult(0);
+  opOperand.set(resolvedValue);
+
+  // Create a new value for the resolved value and subscribe it to propagation
+  // and enforcement.
+  // We possibly don't need to subscribe this since this value has already
+  // reached the top of the lattice and shouldn't do anything else. But it's
```

Comment:
How can we conclude that the value has reached the top of the lattice?

Reply:
An SSA value having a layout, never changes it's layout again. New values may be created with different layout though. So a value having a layout is at top.